unitTests/magicbox/QueryProcessorTest.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+export function QueryProcessorTest() {
+  describe('QueryProcessor', () => {
+    it('does not throw an error when overriding nothing', () => {
+      expect(() => new QueryProcessor().overrideIfProcessing()).not.toThrow();
+    });
+
+    describe('calling processQueries', () => {
+      describe('without values', () => {
+        let queryProcessor: QueryProcessor<string>;
+        let returnedPromise: Promise<IQueryProcessResult<string>>;
+        beforeEach(() => {
+          queryProcessor = new QueryProcessor();
+          returnedPromise = queryProcessor.processQueries([]);
+        });
+
+        it('resolves immediately', async done => {
+          const resolveProcess = jasmine.createSpy('resolveProcess');
+          returnedPromise.then(resolveProcess);
+          await wait();
+          expect(resolveProcess).toHaveBeenCalled();
+          done();
+        });
+
+        it('resolves with Finished status', async done => {
+          expect((await returnedPromise).status).toEqual(ProcessingStatus.Finished);
+          done();
+        });
+
+        it('resolves with an empty array', async done => {
+          expectArrayEqual((await returnedPromise).results, []);
+          done();
+        });
+      });
+
+      const timeBetweenQueries = 10;
+      const resolveTimes = [0, timeBetweenQueries, timeBetweenQueries * 2, timeBetweenQueries * 3];
+      describe(`with values delayed by ${resolveTimes.map(timeout => timeout + 'ms').join(', ')}`, () => {
+        const lastResolveTime = resolveTimes[resolveTimes.length - 1];
+        const stringArraysInResolvedOrder = [['abcd', 'efgh'], ['ijkl'], ['mnop', 'qrst'], ['uvwx']];
+        let stringPromisesInResolvedOrder: (string[] | Promise<string[]>)[];
+        let stringPromisesInParameterOrder: (string[] | Promise<string[]>)[];
+
+        beforeEach(() => {
+          stringPromisesInResolvedOrder = [
+            stringArraysInResolvedOrder[0],
+            wait(resolveTimes[1], stringArraysInResolvedOrder[1]),
+            wait(resolveTimes[2], stringArraysInResolvedOrder[2]),
+            wait(resolveTimes[3], stringArraysInResolvedOrder[3])
+          ];
+          stringPromisesInParameterOrder = [
+            stringPromisesInResolvedOrder[2],
+            stringPromisesInResolvedOrder[1],
+            stringPromisesInResolvedOrder[0],
+            stringPromisesInResolvedOrder[3]
+          ];
+        });
+
+        const whenToOverride = resolveTimes[2];
+        describe(`when overriding after ${whenToOverride}ms`, () => {
+          let queryProcessor: QueryProcessor<string>;
+          let returnedPromise: Promise<IQueryProcessResult<string>>;
+          beforeEach(() => {
+            queryProcessor = new QueryProcessor({ timeout: lastResolveTime + timeBetweenQueries / 2 });
+            returnedPromise = queryProcessor.processQueries(stringPromisesInParameterOrder);
+          });
+
+          function willBeOverriden() {
```

Comment:
Hmm, I see that you have the `it` statements inside this function to avoid repetition, but I think it would be better to repeat in this case for readability and to make it easier to debug/set break-points in the future.

Reply:
Good point, I wasn't sure if I should avoid repetition or not. I'll copy those `it` blocks over.